### PR TITLE
Support native histograms in store-gateway

### PR DESCRIPTION
**Pull request by fpetkovski**
This commit adds support for retrieving native histogram chunks
in Store Gateway.

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

### CODE REVIEW

**Comment by fpetkovski** [comment]
This is currently blocked by https://github.com/thanos-io/thanos/pull/6032.

**Comment by fpetkovski** [comment]
Should be ready for review now.

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -1050,14 +1050,16 @@ func populateChunk(out *storepb.AggrChunk, in chunkenc.Chunk, aggrs []storepb.Ag
...
 	defer hashPool.Put(hasher)
 
-	if in.Encoding() == chunkenc.EncXOR {
+	if in.Encoding() == chunkenc.EncXOR || in.Encoding() == chunkenc.EncHistogram {
 		b, err := save(in.Bytes())
 		if err != nil {
 			return err
 		}
-		out.Raw = &storepb.Chunk{Type: storepb.Chunk_XOR, Data: b, Hash: hashChunk(hasher, b, calculateChecksum)}
+		storeEnc := storepb.Chunk_Encoding(in.Encoding() - 1)
```
Comment (GiedriusS) [inline pkg/store/bucket.go:1058]: I think we generally prefer to avoid variables used just one: https://thanos.io/v0.30/contributing/coding-style-guide.md/#avoid-defining-variables-used-only-once. I suggest moving this into the next line and split it over a few lines instead.

File: pkg/store/storepb/testutil/series.go [hunk 2]
```diff
@@ -143,15 +150,60 @@ func CreateHeadWithSeries(t testing.TB, j int, opts HeadGenOptions) (*tsdb.Head,
 				c.MaxTime = c.MinTime + int64(chEnc.NumSamples()) - 1
 			}
 
+			storeChunkEnc := storepb.Chunk_Encoding(chEnc.Encoding() - 1)
```
Comment (GiedriusS) [inline pkg/store/storepb/testutil/series.go:153]: Same here.

File: pkg/store/storepb/testutil/series.go [hunk 3]
```diff
@@ -256,20 +308,22 @@ func TestServerSeries(t testutil.TB, store storepb.StoreServer, cases ...*Series
...
-					if len(c.ExpectedSeries) > 4 {
+					if len(c.ExpectedSeries) > 1 {
 						for j := range c.ExpectedSeries {
 							testutil.Equals(t, c.ExpectedSeries[j].Labels, srv.SeriesSet[j].Labels, "%v series chunks mismatch", j)
 
 							// Check chunks when it is not a skip chunk query
 							if !c.Req.SkipChunks {
 								if len(c.ExpectedSeries[j].Chunks) > 20 {
-									testutil.Equals(t, len(c.ExpectedSeries[j].Chunks), len(srv.SeriesSet[j].Chunks), "%v series chunks number mismatch", j)
+									require.Equal(t, len(c.ExpectedSeries[j].Chunks), len(srv.SeriesSet[j].Chunks), "%v series chunks number mismatch", j)
```
Comment (GiedriusS) [inline pkg/store/storepb/testutil/series.go:318]: Why the change here? :thinking:

File: pkg/store/storepb/testutil/series.go [hunk 4]
```diff
@@ -256,20 +308,22 @@ func TestServerSeries(t testutil.TB, store storepb.StoreServer, cases ...*Series
...
+								}
+								for ci := range c.ExpectedSeries[j].Chunks {
+									testutil.Equals(t, c.ExpectedSeries[j].Chunks[ci], srv.SeriesSet[j].Chunks[ci], "%v series chunks mismatch %v", j, ci)
 								}
-								testutil.Equals(t, c.ExpectedSeries[j].Chunks, srv.SeriesSet[j].Chunks, "%v series chunks mismatch", j)
 							}
 						}
 					} else {
-						testutil.Equals(t, c.ExpectedSeries, srv.SeriesSet)
+						require.Equal(t, c.ExpectedSeries, srv.SeriesSet)
```
Comment (GiedriusS) [inline pkg/store/storepb/testutil/series.go:326]: Same here.

File: pkg/store/storepb/testutil/series.go [hunk 5]
```diff
@@ -256,20 +308,22 @@ func TestServerSeries(t testutil.TB, store storepb.StoreServer, cases ...*Series
 					}
 
 					// Huge responses can produce unreadable diffs - make it more human readable.
-					if len(c.ExpectedSeries) > 4 {
+					if len(c.ExpectedSeries) > 1 {
```
Comment (GiedriusS) [inline pkg/store/storepb/testutil/series.go:311]: Why this change? If we are doing then perhaps `> 0` makes more sense?

File: pkg/store/storepb/testutil/series.go (hunk 3 above)
Comment (fpetkovski) [reply pkg/store/storepb/testutil/series.go]: I made most of these changes to troubleshoot test failures since `require` produces a nicer diff. They should all be reverted now.

File: pkg/store/bucket.go (hunk 1 above)
Comment (fpetkovski) [reply pkg/store/bucket.go]: Makes sense, updated.

**Review by GiedriusS — APPROVED** [review]
:bow:, just some CI failures left

**Comment by fpetkovski** [comment]
Hm I am not sure if the sonatype-lift comments make sense. Two of them say it cannot import `github.com/prometheus/prometheus/model/histogram` which I don't fully understand, and the third one is related to a crypto function in test code.

File: pkg/testutil/e2eutil/prometheus.go [hunk 6]
```diff
@@ -25,16 +25,19 @@ import (
 	"github.com/go-kit/log"
 	"github.com/oklog/ulid"
 	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/histogram"
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/testutil/e2eutil/prometheus.go:28]: @sonatype-lift ignore

File: pkg/testutil/e2eutil/prometheus.go [hunk 7]
```diff
@@ -462,7 +500,16 @@ func createBlock(
...
 
 				for _, lset := range batch {
-					_, err := app.Append(0, lset, t, rand.Float64())
+					sort.Slice(lset, func(i, j int) bool {
+						return lset[i].Name < lset[j].Name
+					})
+
+					var err error
+					if sampleType == chunkenc.ValFloat {
+						_, err = app.Append(0, lset, t, rand.Float64())
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/testutil/e2eutil/prometheus.go:509]: @sonatype-lift ignore

File: pkg/store/storepb/testutil/series.go [hunk 8]
```diff
@@ -16,6 +16,11 @@ import (
 	"time"
 
 	"github.com/cespare/xxhash"
+	"github.com/prometheus/prometheus/model/histogram"
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/store/storepb/testutil/series.go]: @sonatype-lift ignore
